File: DFS/CourseScheduleII.py
from typing import List

class SolutionDFS:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        self.graph = [[] for _ in range(numCourses)]
        self.visited = [0] * numCourses
        for end, start in prerequisites:
            self.graph[start].append(end)

        self.path = []
        for i in range(numCourses):
            if not self._dfs(i):
                return []
        return self.path

    def _dfs(self, i):
        if self.visited[i] == 1:
            return False
        if self.visited[i] == 2:
            return True

        self.visited[i] = 1
        for nxt in self.graph[i]:
            if not self._dfs(nxt):
                return False
        self.visited[i] = 2
        self.path.append(i)
        return True


numCourses = 4
prerequisites = [[0,1], [1,3], [1,2], [2,3]]
# prerequisites = [[0,1], [3,1], [1,2], [2,3]]
s = SolutionDFS()
res = s.findOrder(numCourses, prerequisites)
print(res)

File: DFS/CourseScheduleII.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using std::vector;

class Solution {
public:
    vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
        vector<vector<int>> graph(numCourses, vector<int> ());
        vector<int> visited(numCourses, 0);
        vector<int> path;
        for (auto p : prerequisites) {
            graph[p[1]].push_back(p[0]);
        }

        for (int i = 0; i < numCourses; i++) {
            if (!dfs(i, graph, visited, path))
                return {};
        }

        std::reverse(path.begin(), path.end());
        return path;
    }

    bool dfs(int i, vector<vector<int>>& graph, vector<int>& visited, vector<int>& path) {
        if (visited[i] == 1) return false;
        if (visited[i] == 2) return true;

        visited[i] = 1;
        for (auto nxt : graph[i]) {
            if (!dfs(nxt, graph, visited, path)) return false;
        }
        visited[i] = 2;
        path.push_back(i);
        return true;
    }
};

int main()
{
    int n = 4;
    vector<vector<int>> pairs {{1,0}, {2,0}, {3,1}, {3,2}};
    Solution s;
    vector<int> res = s.findOrder(n, pairs);
    for (auto n : res)
        std::cout << n << "\n";
    return 0;
}
